lib/consul/async/vault_endpoint.rb

```diff
@@ -0,0 +1,246 @@
+          @response = override_nil_response
+          @error = 'Not initialized yet'
+        else
+          @response_header = http.response_header.nil? ? nil : http.response_header.dup.freeze
+          @response = http.response.nil? || http.response.empty? ? override_nil_response : http.response.dup.freeze
+          @error = http.error.nil? ? nil : http.error.dup.freeze
+        end
+      end
+    end
+    class VaultEndpoint
+      attr_reader :conf, :path, :http_method, :queue, :stats, :last_result, :enforce_json_200, :start_time, :default_value, :query_params
+
+      def initialize(conf, path, http_method = 'GET', enforce_json_200 = true, query_params = {}, default_value = '[]', post_data = {})
+        @conf = conf.create(path)
+        @default_value = default_value
+        @path = path
+        @http_method = http_method
+        @queue = EM::Queue.new
+        @x_consul_index = 0
+        @s_callbacks = []
+        @e_callbacks = []
+        @enforce_json_200 = enforce_json_200
+        @start_time = Time.now.utc
+        @consecutive_errors = 0
+        @query_params = query_params
+        @post_data = post_data
+        @stopping = false
+        @stats = EndPointStats.new
+        @last_result = VaultResult.new(default_value, false, HttpResponse.new(nil), stats, 1)
+        on_response { |result| @stats.on_response result }
+        on_error { |http| @stats.on_error http }
+        _enable_network_debug if conf.debug && conf.debug[:network]
+        fetch
+        queue.push 0
+      end
+
+      def _enable_network_debug
+        on_response do |result|
+          state = result.x_consul_index.to_i < 1 ? '[WARN]' : '[ OK ]'
+          stats = result.stats
+          STDERR.puts "[DBUG]#{state}#{result.modified? ? '[MODFIED]' : '[NO DIFF]'}" \
+          "[s:#{stats.successes},err:#{stats.errors}]" \
+          "[#{stats.body_bytes_human.ljust(8)}][#{stats.bytes_per_sec_human.ljust(9)}]"\
+          " #{path.ljust(48)} idx:#{result.x_consul_index}, next in #{result.retry_in} s"
+        end
+        on_error { |http| STDERR.puts "[ERROR]: #{path}: #{http.error}" }
+      end
+
+      def on_response(&block)
+        @s_callbacks << block
+      end
+
+      def on_error(&block)
+        @e_callbacks << block
+      end
+
+      def ready?
+        @ready
+      end
+
+      def terminate
+        @stopping = true
+      end
+
+      private
+
+      def build_request
+        res = {
+          head: {
+            'Accept' => 'application/json',
+            'X-Vault-Token' => conf.token
+          },
+          query: {},
+          path: path,
+          keepalive: true,
+          callback: method(:on_response)
+        }
+        # if @post_data
+        #   res[:body] = JSON.generate(@post_data)
+        # end
+        @query_params.each_pair do |k, v|
+          res[:query][k] = v
+        end
+        res
+      end
+
+      def get_lease_duration(result)
+        JSON[result]['lease_duration'] || conf.min_duration
```

Comment:
What does this syntax do ? Personally, I hate this JSON[result] ... what does it means?

Reply:
The JSON module has the [] operator as syntax sugar to parse string
```
JSON['{"toto":"tata"}'] == { toto: "tata" } 
```

I'm not attached to it, and the logic is kinda stupid atm, because I parse the response multiple times, this will go in refactoring patch unless required